### PR TITLE
pydantic v2 support

### PR DESCRIPTION
**Pull request by PietroPasotti**
this PR adds pydantic v2 support

### CODE REVIEW

**Review by mmkay — APPROVED** [review]

**Comment by PietroPasotti** [comment]
Come to think of it, given we're effectively breaking BC and possibly introducing incompatibilities if you try and use this together with another charm (lib) that requires pydantic<2, should we actually be bumping the major?
@simskij @mmkay

**Comment by sed-i** [comment]
> Come to think of it, given we're effectively breaking BC and possibly introducing incompatibilities if you try and use this together with another charm (lib) that requires pydantic<2, should we actually be bumping the major? @simskij @mmkay

Yeah, this PR should probably have quite a few tandem PRs.

File: lib/charms/tempo_k8s/v0/tracing.py [hunk 1]
```diff
@@ -121,37 +122,61 @@ class AmbiguousRelationUsageError(TracingError):
...
-            msg = f"failed to validate remote unit databag: {databag}"
+            msg = f"failed to validate databag: {databag}"
             logger.error(msg, exc_info=True)
             raise DataValidationError(msg) from e
 
-    def dump(self, databag: MutableMapping):
-        """Write the contents of this model to Juju databag."""
-        if self._NEST_UNDER:
-            databag[self._NEST_UNDER] = self.json()
+    def dump(self, databag: Optional[MutableMapping] = None, clear: bool = True):
```
Comment (sed-i) [inline lib/charms/tempo_k8s/v0/tracing.py:158]: **Suggestion (non-blocking)**: decouple `dump` from `update`.

Currently, `dump` does two things:
1. dump
2. update (a mutable, "by ref")

I think we should decouple. Decoupling might also obviate _NEST_UNDER and `dump` itself.

Could tackle this in a separate issue.

File: charmcraft.yaml [hunk 2]
```diff
@@ -10,3 +10,8 @@ bases:
       - name: "ubuntu"
         channel: "22.04"
 
+parts:
+  charm:
+    charm-binary-python-packages:
+      - "pydantic>=2"
+      - "opentelemetry-exporter-otlp-proto-grpc==1.17.0"
```
Comment (sed-i) [inline charmcraft.yaml:17]: Would be handy to document deps here:

```suggestion
      # lib/charms/tempo_k8s/v0/tracing.py
      - "pydantic>=2"
      
      - "opentelemetry-exporter-otlp-proto-grpc==1.17.0"
```

**Review by sed-i — APPROVED** [review]

File: lib/charms/tempo_k8s/v0/tracing.py [hunk 3]
```diff
@@ -93,9 +93,9 @@ def __init__(self, *args):
 
 # Increment this PATCH version before using `charmcraft publish-lib` or reset
 # to 0 if you are raising the major API version
-LIBPATCH = 7
+LIBPATCH = 8
```
Comment (mmkay) [inline lib/charms/tempo_k8s/v0/tracing.py:96]: As #52 got merged, you'll need to bump up the version by one more.

```suggestion
LIBPATCH = 9
```

Comment (PietroPasotti) [reply lib/charms/tempo_k8s/v0/tracing.py]: major bumped instead

**Pull request merged**